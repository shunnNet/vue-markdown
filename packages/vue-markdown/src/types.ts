import type { Options as DeepMergeOptions } from 'deepmerge'
import type { Element } from 'hast'
import type { Options as TRehypeOptions } from 'mdast-util-to-hast'
import type { Options } from 'rehype-sanitize'
import type { PluggableList } from 'unified'
import type {
  AllowedComponentProps,
  ComponentCustomProps,
  VNode,
  VNodeArrayChildren,
  VNodeProps,
} from 'vue'

export type Context = {
  listDepth: number
  listOrdered: boolean
  listItemIndex: number
  currentContext?: string
  svg: boolean
}
export type Attributes = Record<string, string>

type TTableProps = {
  /** whether it is in head */
  isHead: boolean
}

type THeadingProps = {
  /** heading level */
  level: number
}

type TListProps = {
  /** depth of the list */
  depth: number
  /** whether it is ordered list */
  ordered: boolean
}

type TCodeProps = {
  /** language name original @example 'language-js' */
  languageOriginal: string

  /** language name @example 'js' */
  language: string

  /** code content */
  content: string

  /** whether it is inline code */
  inline: boolean
}

// https://www.google.com/search?q=record%3Cstring,+any%3E+vs+record%3Cstring,+unknown%3E&sourceid=chrome&ie=UTF-8
export type CustomAttrsObjectResult = Record<string, unknown>

type CustomAttrsFunctionValue<T> = (
  /**
   * hast node
   *
   * Please refer to the source code at the following URL to understand the possible attributes for each element.
   *
   * @see https://github.com/syntax-tree/mdast-util-to-hast/tree/main/lib/handlers
   */
  node: Element,
  /**
   * Properties of the current element.
   *
   * Except for the basic properties provided from hast, it also includes custom properties such as `level`, `ordered`, `depth`, `index` etc.
   */
  combinedAttrs: T | Attributes
) => Record<string, unknown>

type CustomAttrsValue<T extends Record<string, unknown> = Record<string, unknown>>
  = CustomAttrsObjectResult | CustomAttrsFunctionValue<T>

type TBasicHTMLTagNames = keyof Omit<
  HTMLElementTagNameMap,
  'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'ul' | 'ol' | 'li' | 'code' | 'td' | 'th' | 'tr'
>
export type CustomAttrs = {
  [key in TBasicHTMLTagNames]?: CustomAttrsValue // << dynamic properties
} & {
  [key: string]: CustomAttrsValue
    | CustomAttrsValue<THeadingProps>
    | CustomAttrsValue<TListProps>
    | CustomAttrsValue<TCodeProps>
    | CustomAttrsValue<TTableProps>
    | undefined
  ['h1']?: CustomAttrsValue<THeadingProps> // << static properties
  ['h2']?: CustomAttrsValue<THeadingProps>
  ['h3']?: CustomAttrsValue<THeadingProps>
  ['h4']?: CustomAttrsValue<THeadingProps>
  ['h5']?: CustomAttrsValue<THeadingProps>
  ['h6']?: CustomAttrsValue<THeadingProps>
  ['heading']?: CustomAttrsValue<THeadingProps>
  ['ul']?: CustomAttrsValue<TListProps>
  ['ol']?: CustomAttrsValue<TListProps>
  ['list']?: CustomAttrsValue<TListProps>
  ['li']?: CustomAttrsValue<TListProps>
  ['list-item']?: CustomAttrsValue<TListProps>
  ['code']?: CustomAttrsValue<TCodeProps>
  ['inline-code']?: CustomAttrsValue<TCodeProps>
  ['block-code']?: CustomAttrsValue<TCodeProps>
  ['td']?: CustomAttrsValue<TTableProps>
  ['th']?: CustomAttrsValue<TTableProps>
  ['tr']?: CustomAttrsValue<TTableProps>
}

export type AliasList = string[]
export type TagList = AliasList

export type SanitizeOptions = {
  /**
   * Options for `rehype-sanitize`
   *
   * @see https://github.com/rehypejs/rehype-sanitize
   */
  sanitizeOptions?: Options
  /**
   * Options for `deepmerge`
   */
  mergeOptions?: DeepMergeOptions
}

export type TVueMarkdownProps = {
  /**
   * Markdown content
   *
   * @default '''
   */
  markdown: string
  /**
   * You can set custom attributes for each element, such as `href`, `target`, `rel`, `lazyload`, etc.
   *
   * The key is the HTML tag name, and the value can either be an object or a function that returns an object.
   *
   * The value will be passed to Vue's `h` function. You can refer to Vue's official documentation to learn how to configure `h`.
   *
   * @see https://vuejs.org/guide/extras/render-function.html#render-functions-jsx
   *
   * @default {}
   *
   * @example
   * ```ts
   * {
   *  a: { target: '_blank', rel: 'noopener' },
   *  img: { lazyload: true },
   *  h1: (node, combinedAttrs) => {
   *    return { class: ['title', 'mb-2'] }
   *  }
   * }
   * ```
   */
  customAttrs?: CustomAttrs
  /**
   * Remark plugins
   *
   * These plugins will be used between `remark-parse` and `remark-rehype`.
   *
   * @see https://github.com/remarkjs/remark?tab=readme-ov-file#plugins
   *
   * @default []
   */
  remarkPlugins?: PluggableList
  /**
   * rehype plugins
   *
   * These plugins will be used after `remark-rehype` but before `rehype-sanitize`.
   *
   * @see https://github.com/remarkjs/remark-rehype?tab=readme-ov-file#related
   *
   * @default []
   */
  rehypePlugins?: PluggableList
  /**
   * Whether to sanitize the HTML content. (use `rehype-sanitize`)
   *
   * You need disable this option if you want to render `<slot>` in markdown content.
   *
   * @default false
   */
  sanitize?: boolean
  /**
   * Options for `rehype-sanitize`
   *
   * @see https://github.com/rehypejs/rehype-sanitize?tab=readme-ov-file#options
   *
   * @default { allowDangerousHtml: true }
   */
  sanitizeOptions?: SanitizeOptions

  /**
   * Options for `rehype-parse`
   *
   * @see https://github.com/remarkjs/remark-rehype?tab=readme-ov-file#options
   *
   * @default {}
   */
  rehypeOptions?: Omit<TRehypeOptions, 'file'>
}

/**
 * Typed version of the `VueMarkdown` component.
 *
 * Copy from vue-router
 */
export type TVueMarkdown = {
  new(): {
    $props: AllowedComponentProps
      & ComponentCustomProps
      & VNodeProps
      & TVueMarkdownProps

    $slots: TBaseSlots & {
      /**
       * Customize `<h1>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h1']?: THeadingSlot
      /**
       * Customize `<h2>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h2']?: THeadingSlot
      /**
       * Customize `<h3>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h3']?: THeadingSlot
      /**
       * Customize `<h4>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h4']?: THeadingSlot
      /**
       * Customize `<h5>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h5']?: THeadingSlot
      /**
       * Customize `<h6>` tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['h6']?: THeadingSlot
      /**
       * Customize `<h1>` ~ `<h6>`  tag
       * @scope `level` heading level
       * @scope `children` Functional component, child elements.
       */
      ['heading']?: THeadingSlot

      /**
       * Customize inline and block code.
       * @scope `languageOriginal` language name original
       * @scope `language` language name
       * @scope `content` code content
       * @scope `inline` whether it is inline code
       * @scope `children` Functional component, child elements.
       */
      ['code']?: TCodeSlot
      /**
       * Customize inline code.
       * @scope `languageOriginal` language name original
       * @scope `language` language name
       * @scope `content` code content
       * @scope `inline` whether it is inline code
       * @scope `children` Functional component, child elements.
       */
      ['inline-code']?: TCodeSlot
      /**
       * Customize block code.
       * @scope `languageOriginal` language name original
       * @scope `language` language name
       * @scope `content` code content
       * @scope `inline` whether it is inline code
       * @scope `children` Functional component, child elements.
       */
      ['block-code']?: TCodeSlot

      /**
       * Customize unordered list
       * @scope `depth` depth of the list
       * @scope `ordered` whether it is ordered list
       * @scope `children` Functional component, child elements.
       */
      ['ul']?: TListSlot

      /**
       * Customize ordered list
       * @scope `depth` depth of the list
       * @scope `ordered` whether it is ordered list
       * @scope `children` Functional component, child elements.
       */
      ['ol']?: TListSlot
      /**
       * Customize ordered and unordered list
       * @scope `depth` depth of the list
       * @scope `ordered` whether it is ordered list
       * @scope `children` Functional component, child elements.
       */
      ['list']?: TListSlot

      /**
       * Customize list item
       * @scope `depth` depth of the list
       * @scope `ordered` whether it is ordered list
       * @scope `index` index of the list item
       * @scope `children` Functional component, child elements.
       */
      ['li']?: TListSlot

      /**
       * Customize list item
       * @scope `depth` depth of the list
       * @scope `ordered` whether it is ordered list
       * @scope `index` index of the list item
       * @scope `children` Functional component, child elements.
       */
      ['list-item']?: TListSlot

      /**
       * Customize table element: td
       *
       * @scope `isHead` whether it is in head
       * @scope `children` Functional component, child elements.
       */
      ['td']?: TTableElementSlot

      /**
       * Customize table element: th
       *
       * @scope `isHead` whether it is in head
       * @scope `children` Functional component, child elements.
       */
      ['th']?: TTableElementSlot

      /**
       * Customize table element: tr
       *
       * @scope `isHead` whether it is in head
       * @scope `children` Functional component, child elements.
       */
      ['tr']?: TTableElementSlot
    }
  }
}

type TTableElementSlot = TCustomSlot<TTableProps>
type TListSlot = TCustomSlot<TListProps>
type THeadingSlot = TCustomSlot<THeadingProps>
type TCodeSlot = TCustomSlot<TCodeProps>

type HtmlTagNames = keyof HTMLElementTagNameMap
type TBaseSlots = {
  // An index signature parameter type cannot be a literal type or generic type. Consider using a mapped object type instead.ts(1337)
  // [key: HtmlTagNames]: (scope: Record<string, any>) => VNode[] | VNode
  [key in HtmlTagNames]?: (scope: TBaseSlotScope) => VNode[] | VNode
} & {
  [key: string]: (scope: TBaseSlotScope) => VNode[] | VNode
}

type TBaseSlotScope = TElementChild & Attributes
type TElementChild = {
  /** Functional component, child elements.  */
  children: () => VNodeArrayChildren
}

type TCustomSlot<T> = (
  scope: TBaseSlotScope & T
) => VNode[] | VNode
