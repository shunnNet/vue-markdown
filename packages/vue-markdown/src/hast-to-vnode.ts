import type { Element, Root, RootContent, Text } from 'hast'
import type { MaybeRefOrGetter, Slots, VNode, VNodeArrayChildren } from 'vue'
import type { AliasList, Attributes, Context, CustomAttrs, CustomAttrsObjectResult } from './types'
import { find, html, svg } from 'property-information'
import { h, toValue } from 'vue'

export function render(
  hast: Root,
  attrs: Record<string, unknown>,
  slots?: Slots,
  customAttrs?: MaybeRefOrGetter<CustomAttrs>,
): VNode {
  return h(
    'div',
    attrs,
    renderChildren(
      hast.children,
      { listDepth: -1, listOrdered: false, listItemIndex: -1, svg: false },
      hast,
      slots ?? {},
      toValue(customAttrs) ?? {},
    ),
  )
}

export function renderChildren(
  nodeList: (RootContent | Root)[],
  ctx: Context,
  parent: Element | Root,
  slots: Slots,
  customAttrs: CustomAttrs,
): VNodeArrayChildren {
  const keyCounter: {
    [key: string]: number
  } = {}

  return nodeList.map((node) => {
    switch (node.type) {
      case 'text':
        return node.value
      case 'raw':
        return node.value
      // Note: targeting libs like `@shikijs/rehype` which return `root` node.
      // This may be abnormal. keep track of this.
      case 'root':
        return renderChildren(node.children, ctx, parent, slots, customAttrs)
      case 'element': {
        const { attrs, context, aliasList, vnodeProps } = getVNodeInfos(node, parent, ctx, keyCounter, customAttrs)
        for (let i = aliasList.length - 1; i >= 0; i--) {
          const targetSlot = slots[aliasList[i]]
          if (typeof targetSlot === 'function') {
            return targetSlot({
              ...vnodeProps,
              ...attrs,
              children: () => renderChildren(node.children, context, node, slots, customAttrs),
            })
          }
        }

        return h(node.tagName, attrs, renderChildren(node.children, context, node, slots, customAttrs))
      }
      default:
        return null
    }
  })
}

export function getVNodeInfos(
  node: RootContent,
  parent: Element | Root,
  context: Context,
  keyCounter: Record<string, number>,
  customAttrs: CustomAttrs,
): {
    attrs: Record<string, unknown>
    context: Context
    aliasList: AliasList
    vnodeProps: Record<string, any>
  } {
  const aliasList: AliasList = []

  let attrs: Record<string, unknown> = {}
  const vnodeProps: Record<string, any> = {}
  const ctx = { ...context }

  if (node.type === 'element') {
    aliasList.push(node.tagName)
    keyCounter[node.tagName] = node.tagName in keyCounter ? keyCounter[node.tagName] + 1 : 0
    vnodeProps.key = `${node.tagName}-${keyCounter[node.tagName]}`
    node.properties = node.properties || {}

    if (node.tagName === 'svg') {
      ctx.svg = true
    }

    attrs = Object.entries(node.properties).reduce<Record<string, any>>((acc, [hastKey, value]) => {
      const attrInfo = find(ctx.svg ? svg : html, hastKey)
      acc[attrInfo.attribute] = value

      return acc
    }, {})

    switch (node.tagName) {
      case 'h1':
      case 'h2':
      case 'h3':
      case 'h4':
      case 'h5':
      case 'h6':
        vnodeProps.level = Number.parseFloat(node.tagName.slice(1))
        aliasList.push('heading')
        break
        // TODO: maybe use <pre> instead for customizing from <pre> not <code> ?
      case 'code':
        vnodeProps.languageOriginal = Array.isArray(attrs.class)
          ? attrs.class.find(cls => cls.startsWith('language-'))
          : ''
        vnodeProps.language = vnodeProps.languageOriginal ? vnodeProps.languageOriginal.replace('language-', '') : ''
        vnodeProps.inline = 'tagName' in parent && parent.tagName !== 'pre'

        // when tagName is code, it definitely has children and the first child is text
        // https://github.com/syntax-tree/mdast-util-to-hast/blob/main/lib/handlers/code.js
        vnodeProps.content = (node.children[0] as unknown as Text).value

        aliasList.push(vnodeProps.inline ? 'inline-code' : 'block-code')
        break
      case 'thead':
      case 'tbody':
        ctx.currentContext = node.tagName
        break
      case 'td':
      case 'th':
      case 'tr':
        vnodeProps.isHead = context.currentContext === 'thead'
        break

      case 'ul':
      case 'ol':
        ctx.listDepth = context.listDepth + 1
        ctx.listOrdered = node.tagName === 'ol'
        ctx.listItemIndex = -1
        vnodeProps.ordered = ctx.listOrdered
        vnodeProps.depth = ctx.listDepth

        aliasList.push('list')
        break

      case 'li':
        ctx.listItemIndex++

        vnodeProps.ordered = ctx.listOrdered
        vnodeProps.depth = ctx.listDepth
        vnodeProps.index = ctx.listItemIndex
        aliasList.push('list-item')

        break
      case 'slot':
        if (typeof node.properties['slot-name'] === 'string') {
          aliasList.push(`s-${node.properties['slot-name']}`)
          delete node.properties['slot-name']
        }
        break
      default:
        break
    }

    attrs = computeAttrs(
      node,
      aliasList,
      vnodeProps,
      { ...attrs } as Attributes, // TODO: fix this
      customAttrs,
    )
  }

  return {
    attrs,
    context: ctx,
    aliasList,
    vnodeProps,
  }
}

/**
 * TODO:
 * @param node - hast node
 * @param aliasList - html tag list. The earlier alias has higher priority. ?
 * @param vnodeProps - vnode props
 * @param attrs - attrs
 * @param customAttrs - custom attrs object
 * @returns attrs
 */
function computeAttrs(
  node: Element,
  aliasList: AliasList,
  vnodeProps: Record<string, any>,
  attrs: Attributes,
  customAttrs: CustomAttrs,
): CustomAttrsObjectResult {
  const result: CustomAttrsObjectResult = {
    ...attrs,
  }
  for (let i = aliasList.length - 1; i >= 0; i--) {
    const name = aliasList[i]
    // console.log(Object.keys(customAttrs))
    if (name in customAttrs) {
      const customAttr = customAttrs[name]
      return {
        ...result,
        ...(typeof customAttr === 'function' ? customAttr(node, { ...attrs, ...vnodeProps }) : customAttr),
      }
    }
  }
  return result
}
