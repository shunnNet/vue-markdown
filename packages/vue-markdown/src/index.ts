export { VueMarkdown, VueMarkdownAsync } from './components'
export { getVNodeInfos, render, renderChildren } from './hast-to-vnode'
export type { CustomAttrs, TVueMarkdown, TVueMarkdownProps } from './types'
export { createProcessor, useMarkdownProcessor } from './useProcessor'
