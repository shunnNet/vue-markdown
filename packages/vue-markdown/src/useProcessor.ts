import type { Root } from 'hast'
import type { Root as MdastRoot } from 'mdast'
import type { Options as TRehypeOptions } from 'mdast-util-to-hast'
import type { ComputedRef, MaybeRefOrGetter } from 'vue'
import type { SanitizeOptions } from './types'
import deepmerge from 'deepmerge'
import rehypeSanitize, { defaultSchema } from 'rehype-sanitize'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { type PluggableList, type Processor, unified } from 'unified'
import { computed, toValue } from 'vue'

export type TUseMarkdownProcessorOptions = {
  remarkPlugins?: MaybeRefOrGetter<PluggableList>
  rehypePlugins?: MaybeRefOrGetter<PluggableList>
  rehypeOptions?: MaybeRefOrGetter<Omit<TRehypeOptions, 'file'>>
  sanitize?: MaybeRefOrGetter<boolean>
  sanitizeOptions?: MaybeRefOrGetter<SanitizeOptions>
}
export function useMarkdownProcessor(options?: TUseMarkdownProcessorOptions): { processor: ComputedRef<Processor<MdastRoot, MdastRoot, Root, undefined, undefined>> } {
  const processor = computed(() => {
    return createProcessor({
      prePlugins: [remarkParse, ...(toValue(options?.remarkPlugins) ?? [])],
      rehypePlugins: toValue(options?.rehypePlugins),
      rehypeOptions: toValue(options?.rehypeOptions),
      sanitize: toValue(options?.sanitize),
      sanitizeOptions: toValue(options?.sanitizeOptions),
    })
  })
  return { processor }
}

export function createProcessor(options?: {
  prePlugins?: PluggableList
  rehypePlugins?: PluggableList
  rehypeOptions?: Omit<TRehypeOptions, 'file'>
  sanitize?: boolean
  sanitizeOptions?: SanitizeOptions
  // TODO: fix types
}): Processor<any, any, any, any, any> {
  return unified()
    .use(options?.prePlugins ?? [])
    .use(remarkRehype, { allowDangerousHtml: true, ...(options?.rehypeOptions || {}) })
    .use(options?.rehypePlugins ?? [])
    .use(
      options?.sanitize
        ? [
            [
              rehypeSanitize,
              deepmerge(
                defaultSchema,
                options?.sanitizeOptions?.sanitizeOptions || {},
                options?.sanitizeOptions?.mergeOptions || {},
              ),
            ],
          ]
        : [],
    )
}
