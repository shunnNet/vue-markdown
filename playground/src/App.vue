<script setup lang="ts">
import {VueMarkdown} from "@crazydos/vue-markdown"
import type {CustomAttrs} from "@crazydos/vue-markdown"
import CodeBlock from "./components/CodeBlock.vue"
import { ref }  from "vue"
import remarkGfm from "remark-gfm"
import rehypeRaw from "rehype-raw"
import remarkMath from "remark-math"
import rehypeKatex from "rehype-katex"
import remarkToc from "remark-toc"

const markdown = ref(`# Markdown Test

## Structure

## Autolink literals

www.example.com, https://example.com, and contact@example.com.

## Footnote

A note[^1]

[^1]: Big note.

## Strikethrough

~one~ or ~~two~~ tildes.

## Table

| a | b  |  c |  d  |
| - | - | - | - |
| 1 | 2 | 3 | 4 |

## Tasklist

* [ ] to do
* [x] done

## Raw
<div class="note">

A mix of *markdown* and <em>HTML</em>.

</div>

## Math
Lift($$L$$) can be determined by Lift Coefficient ($$C_L$$) like the following
equation.
$$
L = \\frac{1}{2} \\rho v^2 S C_L
$$

## Vue Slot
<div>
<slot slot-name="header" value="123">I am slot</slot>
</div>

## Svg
<svg viewBox="0 0 800 400" xmlns="http://www.w3.org/2000/svg">
  <rect x="50" y="150" width="100" height="100" fill="#f0f0f0" stroke="#333" stroke-width="5"/>
  <text x="100" y="200" text-anchor="middle" font-family="Arial" font-size="14">Input</text>
</svg>
`)

// const customAttrs: CustomAttrs = {
//   heading: (node, combinedAttrs) => {
//     // if (node.children[0]?.type === 'text') {
//     //   return {
//     //     id: node.children[0].value.toLowerCase().replace(/\s/g, '-'),
//     //     ...combinedAttrs
//     //   }
//     // }

//     return { level: 1, onClick: () => { console.log(123) } }
//   }
// }
const customAttrs: CustomAttrs = {
  h1: { 'class': ["heading"] },
  h2: { 'class': ["heading"] },
  a: (node, combinedAttrs) => {
    if(
      typeof node.properties.href === 'string' &&
      node.properties.href.startsWith('https://www.google.com')
    ){
      return { target: '_blank', rel: "noopener noreferrer"}
    } else {
      return {}
    }
  }
}

const preset = {
  plugins: [
    remarkGfm,
    remarkMath,
  ]
}

</script>

<template>
  <div>

    <VueMarkdown
      :markdown="markdown"
      :custom-attrs="customAttrs"
      :remark-plugins="[preset, [remarkToc, {heading: 'structure'}]]"
      :rehype-plugins="[rehypeRaw, rehypeKatex]"
    >
      <template v-slot:s-header="{ children: Children, ...attrs }">
        <p v-bind="attrs">
          <component :is="Children"  />
        </p>
      </template>

      <template #code="{ children, ...props}">
        <CodeBlock :code="props.content" />
      </template>

    </VueMarkdown>
  </div>
</template>

<style>

</style>
