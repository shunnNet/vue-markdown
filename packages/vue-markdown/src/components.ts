import type { Root } from 'hast'
import type { Options as TRehypeOptions } from 'mdast-util-to-hast'
import type { PluggableList } from 'unified'
import type { PropType } from 'vue'

import type { CustomAttrs, SanitizeOptions, TVueMarkdown } from './types'
import {
  defineComponent,
  toRefs,
} from 'vue'
import { render } from './hast-to-vnode'
import { useMarkdownProcessor } from './useProcessor'

export type { CustomAttrs, SanitizeOptions, TVueMarkdown }

const sharedProps = {
  markdown: {
    type: String as PropType<string>,
    default: '',
  },
  customAttrs: {
    type: Object as PropType<CustomAttrs>,
    default: () => ({}),
  },
  remarkPlugins: {
    type: Array as PropType<PluggableList>,
    default: () => [],
  },
  rehypePlugins: {
    type: Array as PropType<PluggableList>,
    default: () => [],
  },
  rehypeOptions: {
    type: Object as PropType<Omit<TRehypeOptions, 'file'>>,
    default: () => ({}),
  },
  sanitize: {
    type: Boolean,
    default: false,
  },
  sanitizeOptions: {
    type: Object as PropType<SanitizeOptions>,
    default: () => ({}),
  },
}
const vueMarkdownImpl = defineComponent({
  name: 'VueMarkdown',
  props: sharedProps,
  setup(props, { slots, attrs }) {
    const { markdown, remarkPlugins, rehypePlugins, rehypeOptions, sanitize, sanitizeOptions, customAttrs } = toRefs(props)
    const { processor } = useMarkdownProcessor({
      remarkPlugins,
      rehypePlugins,
      rehypeOptions,
      sanitize,
      sanitizeOptions,
    })

    return () => {
      const mdast = processor.value.parse(markdown.value)
      const hast = processor.value.runSync(mdast) as Root
      return render(hast, attrs, slots, customAttrs.value)
    }
  },
})

// export the public type for h/tsx inference
// also to avoid inline import() in generated d.ts files
export const VueMarkdown: TVueMarkdown = vueMarkdownImpl as any
