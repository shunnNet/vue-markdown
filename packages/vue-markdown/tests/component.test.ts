import { mount } from '@vue/test-utils'
import rehypeHighlight from 'rehype-highlight'
import rehypeRaw from 'rehype-raw'
import remarkGfm from 'remark-gfm'
import { describe, expect, it } from 'vitest'
import { VueMarkdown } from '../src/old'

describe('vueMarkdown', () => {
  it('should render markdown and HTML correctly', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

hello world

<div>test</div>

<!-- test -->
`,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with remark plugin correctly', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

- [ ] todo 1
- [x] todo 2

| header 1 | header 2 |
|----------|----------|
| cell 1   | cell 2   |
`,
        remarkPlugins: [remarkGfm],
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with rehype plugin correctly', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

\`\`\`js
const hello = 'world'
console.log(hello)
\`\`\`

`,
        rehypePlugins: [rehypeHighlight],
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with rehypeOptions correctly', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

A note[^1]

[^1]: Big note.
`,
        remarkPlugins: [remarkGfm],
        rehypeOptions: {
          clobberPrefix: 'test-',
        },
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render raw HTML when sanitize is false', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

<div>valid div</div>
<script>alert('invalid script')</script>
<iframe src="invalid-iframe"></iframe>
`,
        rehypePlugins: [rehypeRaw],
        sanitize: false,
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with sanitizeOptions1 correctly', () => {
    const wrapper = mount(VueMarkdown, {
      props: {
        markdown: `
# title

<div>valid div</div>
<script>alert('invalid script')</script>
<iframe src="invalid-iframe"></iframe>
`,
        rehypePlugins: [rehypeRaw],
        sanitize: true,
        sanitizeOptions: {
          sanitizeOptions: {
            tagNames: [],
          },
          // mergeOptions: Optional. Internally, we use the `deepmerge` package to combine `defaultSchema` and `sanitizeOptions`. You can adjust the merging behavior in `mergeOptions`.
          mergeOptions: {
            arrayMerge: (_, source) => {
              return source
            },
          },
        },
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
