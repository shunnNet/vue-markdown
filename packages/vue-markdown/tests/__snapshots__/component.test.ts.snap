// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vueMarkdown > should render markdown and HTML correctly 1`] = `
"<div>
  <h1>title</h1>
  <p>hello world</p>
  <!---->
  <!---->
</div>"
`;

exports[`vueMarkdown > should render raw HTML when sanitize is false 1`] = `
"<div>
  <h1>title</h1>
  <div>valid div</div>
  <script>
    alert('invalid script')
  </script>
  <iframe src="invalid-iframe"></iframe>
</div>"
`;

exports[`vueMarkdown > should render with rehype plugin correctly 1`] = `
"<div>
  <h1>title</h1>
  <pre><code class="hljs language-js"><span class="hljs-keyword">const</span> hello = <span class="hljs-string">'world'</span>
<span class="hljs-variable language_">console</span>.<span class="hljs-title function_">log</span>(hello)
</code></pre>
</div>"
`;

exports[`vueMarkdown > should render with rehypeOptions correctly 1`] = `
"<div>
  <h1>title</h1>
  <p>A note<sup><a href="#test-fn-1" id="test-fnref-1" data-footnote-ref="true" aria-describedby="footnote-label">1</a></sup></p>
  <section data-footnotes="true" class="footnotes">
    <h2 class="sr-only" id="footnote-label">Footnotes</h2>
    <ol>
      <li id="test-fn-1">
        <p>Big note. <a href="#test-fnref-1" data-footnote-backref="" aria-label="Back to reference 1" class="data-footnote-backref">â†©</a></p>
      </li>
    </ol>
  </section>
</div>"
`;

exports[`vueMarkdown > should render with remark plugin correctly 1`] = `
"<div>
  <h1>title</h1>
  <ul class="contains-task-list">
    <li class="task-list-item"><input type="checkbox" disabled=""> todo 1</li>
    <li class="task-list-item"><input type="checkbox" checked="" disabled=""> todo 2</li>
  </ul>
  <table>
    <thead>
      <tr>
        <th>header 1</th>
        <th>header 2</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>cell 1</td>
        <td>cell 2</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`vueMarkdown > should render with sanitizeOptions correctly 1`] = `
"<div>title
  valid div

</div>"
`;

exports[`vueMarkdown > should render with sanitizeOptions1 correctly 1`] = `
"<div>title
  valid div

</div>"
`;
